### PR TITLE
[type:feature] Adding a Motan plugin 、Spring Cloud and WebSocket plugin to shenyu ingress-controller

### LINKED ISSUE #5053: [Task] Adding a Motan plugin 、Spring Cloud and WebSocket plugin to shenyu ingress-controller
### Description

see https://github.com/apache/shenyu/issues/4581 https://github.com/apache/shenyu/issues/4438

### Task List

- [x] Adding a Motan plugin to shenyu ingress-controller
- [x]  Support for Motan TLS configuration
- [x] Adding a Spring Cloud plugin to shenyu ingress-controller
- [x] Support for Spring Cloud TLS configuration
- [x] Adding a WebSocket plugin to shenyu ingress-controller
- [x] Adding a WebSocket plugin to shenyu ingress-controller

### PR DESCRIPTION
**Pull request by runqi-zhao**
fix #5053 

Make sure that:

- [x] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

**Review by JooKS-me — CHANGES_REQUESTED** [review]

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DubboIngressParser.java [hunk 1]
```diff
@@ -292,7 +292,7 @@ private List<DubboUpstream> parseUpstream(final V1IngressBackend backend, final
                         DubboUpstream upstream = DubboUpstream.builder()
                                 .upstreamUrl(clusterIP + ":" + defaultPort)
                                 .weight(100)
-                                .protocol("http://")
+                                .protocol("dubbo://")
```
Comment (moremind) [inline shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DubboIngressParser.java:295]: can the user config protocol?

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/MotanIngressParser.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        ConditionData hostCondition = null;
+        if (Objects.nonNull(ingressRule.getHost())) {
+            hostCondition = new ConditionData();
+            hostCondition.setParamType(ParamTypeEnum.URI.getName());
+            hostCondition.setOperator(OperatorEnum.EQ.getAlias());
+            hostCondition.setParamValue(ingressRule.getHost());
+        }
+        if (Objects.nonNull(ingressRule.getHttp())) {
+            List<V1HTTPIngressPath> paths = ingressRule.getHttp().getPaths();
+            if (paths != null) {
```
Comment (moremind) [inline shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/MotanIngressParser.java:152]: use Objects.nonNull()

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/MotanIngressParser.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                            .pluginName(PluginEnum.MOTAN.getName())
+                            .name(path.getPath())
+                            .matchMode(MatchModeEnum.AND.getCode())
+                            .type(SelectorTypeEnum.CUSTOM_FLOW.getCode())
+                            .enabled(true)
+                            .logged(false)
+                            .continued(true)
+                            .conditionList(conditionList).build();
+
+                    RuleData ruleData = RuleData.builder()
```
Comment (moremind) [inline shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/MotanIngressParser.java:197]: can the user config these?

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/reconciler/IngressReconciler.java [hunk 4]
```diff
@@ -439,4 +433,24 @@ private void addNewIngressConfigToShenyu(final V1Ingress v1Ingress, final CoreV1
...
         }
     }
+
+    private String getPluginName(final V1Ingress ingress) {
+        String pluginName;
+        String pluginDubboEnabled = ingress.getMetadata().getAnnotations().get(IngressConstants.PLUGIN_DUBBO_ENABLED);
+        String pluginMotanEnabled = ingress.getMetadata().getAnnotations().get(IngressConstants.PLUGIN_MOTAN_ENABLED);
+        String pluginSpringCloudEnabled = ingress.getMetadata().getAnnotations().get(IngressConstants.PLUGIN_SPRING_CLOUD_ENABLED);
+        String pluginWebSocketEnabled = ingress.getMetadata().getAnnotations().get(IngressConstants.PLUGIN_WEB_SOCKET_ENABLED);
+        if ("true".equals(pluginDubboEnabled)) {
```
Comment (moremind) [inline shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/reconciler/IngressReconciler.java:443]: Boolean.TRUE

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DubboIngressParser.java (hunk 1 above)
Comment (runqi-zhao) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DubboIngressParser.java:295]: Users cannot set their own protocols for the time being.

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/MotanIngressParser.java (hunk 3 above)
Comment (runqi-zhao) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/MotanIngressParser.java:197]: yes，user can config from yml

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/reconciler/IngressReconciler.java (hunk 4 above)
Comment (JooKS-me) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/reconciler/IngressReconciler.java:443]: ping @runqi-zhao

Comment (runqi-zhao) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/reconciler/IngressReconciler.java:443]: Ok I'm going to modify it using BOLEAN.TRUE.toString

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java [hunk 5]
```diff
@@ -290,15 +292,15 @@ private List<DivideUpstream> parseUpstream(final V1IngressBackend backend, final
...
                     for (V1EndpointAddress address : addresses) {
                         String upstreamIp = address.getIp();
                         String defaultPort = parsePort(backend.getService());
                         if (Objects.nonNull(defaultPort)) {
                             DivideUpstream upstream = new DivideUpstream();
                             upstream.setUpstreamUrl(upstreamIp + ":" + defaultPort);
                             upstream.setWeight(100);
-                            // TODO support config protocol in annotation
-                            upstream.setProtocol("http://");
+                            upstream.setProtocol(Objects.isNull(protocol) ? "http://" : protocol.split(",")[i++]);
```
Comment (JooKS-me) [inline shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java:303]: Why is an array?

Comment (runqi-zhao) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java:303]: In the yaml file, we need to set up several different Upstreams at the same time, so we will set up the yaml file in the following form:
```yaml
metadata.
  name: ...
  annotations.
    yourAnnotationKey: ... 
      - value1
      - value2
      - value3
```
Then read in [value1,value2,value3], so it's good to use an array to iterate through each of these values

Comment (JooKS-me) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java:302]: The split is used once every cycle, which is too inefficient, and the split should be placed outside the cycle.

Comment (runqi-zhao) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java:302]: done

> The split is used once every cycle, which is too inefficient, and the split should be placed outside the cycle.

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java [hunk 6]
```diff
@@ -275,13 +275,17 @@ private String parsePort(final V1IngressServiceBackend service) {
...
-    private List<DivideUpstream> parseUpstream(final V1IngressBackend backend, final String namespace) {
+    private List<DivideUpstream> parseUpstream(final V1IngressBackend backend, final String namespace, final Map<String, String> annotations) {
         List<DivideUpstream> upstreamList = new ArrayList<>();
         if (Objects.nonNull(backend) && Objects.nonNull(backend.getService()) && Objects.nonNull(backend.getService().getName())) {
             String serviceName = backend.getService().getName();
             // shenyu routes directly to the container
             V1Endpoints v1Endpoints = endpointsLister.namespace(namespace).get(serviceName);
             List<V1EndpointSubset> subsets = v1Endpoints.getSubsets();
+            String[] protocol = null;
+            if (Objects.isNull(annotations) && annotations.containsKey(IngressConstants.UPSTREAMS_PROTOCOL_ANNOTATION_KEY)) {
```
Comment (JooKS-me) [inline shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java:286]: maybe isNonNull?

**Review by JooKS-me — CHANGES_REQUESTED** [review]

File: shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java (hunk 6 above)
Comment (runqi-zhao) [reply shenyu-kubernetes-controller/src/main/java/org/apache/shenyu/k8s/parser/DivideIngressParser.java]: done

**Review by JooKS-me — APPROVED** [review]

**Review by moremind — APPROVED** [review]

**Comment by JooKS-me** [comment]
@runqi-zhao The next step is to add integration tests first.

**Comment by runqi-zhao** [comment]
> @runqi-zhao The next step is to add integration tests first.

ok

**Pull request merged**